### PR TITLE
Rework Free to use Coyoneda behind the scenes

### PR DESCRIPTION
**Pull request by adelbertc**
Taken from
https://github.com/scalaz/scalaz/commit/783eb0ed7f6aec56d6245e9b2b9bac97aab74d22#diff-afd5b6310425d773ca3459ebb17ed594

### CODE REVIEW

**Comment by adelbertc** [comment]
Review by @mandubian

**Comment by adelbertc** [comment]
Strange Travis errors, `free` tests seems to be hanging (same behavior when I try locally): https://travis-ci.org/non/cats/jobs/76387437

**Comment by non** [comment]
So, I haven't studied it as closely as I need to, but does this change break the machinery that reassociates left-associated binds to the right?

**Comment by adelbertc** [comment]
@non I believe that's covered in the `foldMap` (and `step`) case.

**Comment by non** [comment]
Aha, I see. I was used to seeing it happen during `flatMap`, but yes, that looks correct.

File: free/src/main/scala/cats/free/Free.scala [hunk 1]
```diff
@@ -3,63 +3,41 @@ package free
...
-   */
-  def runFC[S[_], M[_], A](fa: FreeC[S, A])(f: S ~> M)(implicit M: Monad[M]): M[A] =
-    fa.foldMap[M](new (Coyoneda[S, ?] ~> M) {
-      def apply[B](ca: Coyoneda[S, B]): M[B] = M.map(f(ca.fi))(ca.k)
-    })
+  /** Lift a pure value into Free */
+  def pure[S[_], A](a: A): Free[S, A] = Pure(a)
 
   /**
    * `Free[S, ?]` has a monad if `S` has a `Functor`.
```
Comment (fthomas) [inline free/src/main/scala/cats/free/Free.scala]: This is obviously not true anymore.

**Comment by adelbertc** [comment]
......Travis jobs aren't hanging, I introduced an infinite loop somewhere. Looking into it.

**Comment by mandubian** [comment]
I think I need to think a bit more about what Runar has done before answering to this!
It's really interesting actually!

**Comment by non** [comment]
:+1:

**Comment by mandubian** [comment]
as usual Runar's work is great (and yours too) and I say :+1: for this

**Comment by stew** [comment]
:+1:

**Pull request merged**